### DISCUSSION TITLE
Integrate multi-file context

### DISCUSSION THREAD

**Discussion by nenkoru** [huggingface/llm.nvim#24] [Ideas]
I was wondering that it would be nice-to-have multi-file context by integration of usage of local vector storage.
For instance I would like to have a multi-file context for current completion.
Consider this case:

```utils.py
def get_time(): return 101
```
```main.py
import utils
print(utils.get_t
```
I would like the plugin to complete the rest of the function call for me.
For now we are only getting the context before and after a cursor.
But what if we could get a search within a vector store and embed those pieces within a context? Particularly in `before` window.

**Comment by nenkoru** [comment]
I believe to make it less dependent on a specific vector db or a piece of software that could provide those chunks of code to inject into the context window, we could make a unified REST API interface. So that you just need to implement and run a local(or a remote) web-server which would provide a simple `/search` functionality for the plugin.

**Comment by McPatate** [comment]
I was thinking about how making the context of the prompt more relevant to the actual piece of code being worked on.
My initial idea was to look into working with treesitter to fetch the different functions being called in the current scope of the cursor position. I think having a vector db filled with the different symbols and pieces of code may be simpler to use and while not being perfectly relevant might bring in additional context based on proximity that would not be present with a treesitter approach.
Here's a [good resource](https://thakkarparth007.github.io/copilot-explorer/posts/copilot-internals.html) on how the Copilot plugin internals work - yet another approach to building a good prompt.

**Reply by nenkoru** [reply]
Alright, I have tried replicating the same example of a prompt provided in that article - and it worked.

https://gist.github.com/nenkoru/fc6f0d517655a4cded9ac321611fa9fa

Try to autocomplete print statement. Like remove the invocation of a function within print statement and see what happens.
Also you could try asking for a list of functions in utils.timezone module. Tried with WizardCoder 1.0. FIM enabled.

**Reply by nenkoru** [reply]
Now what's left is to find the way of how to obtain those additional snippets, as well managing the maximum context of a model that is being used.

**Reply by McPatate** [reply]
> Now what's left is to find the way of how to obtain those additional snippets

The hard part 😄

> as well managing the maximum context of a model that is being used

I think for that at the moment we can rely on the config you added and let users set  the correct values.

We're going to have to think about setting values in the before or after part when `fim.enabled == true`. Maybe instead of having two configuration fields before/after length, we can have only one `context_length` parameter and devise the appropriate strategy to create the prompt.

**Reply by McPatate** [reply]
It also works nicely with `starcoder*`

**Comment by McPatate** [comment]
> I would like the plugin to complete the rest of the function call for me.

Building a `hfcc-cmp` was on my todo-list, I haven't looked into how [copilot-cmp](https://github.com/zbirenbaum/copilot-cmp) or [cmp-tabnine](https://github.com/tzachar/cmp-tabnine) are built yet, so I don't know if the vector db approach is overkill, meaning there may be a simpler solution.

**Reply by nenkoru** [reply]
As far as I can understand, cmp-tabnine for instance doesn't have an ability to include files. 
It just includes lines of the current file the same way we do.
(I actually think that it's better in terms of verbosity of how it basically should work vs. amount of characters in the current implementation of hfcc.nvim)

**Reply by nenkoru** [reply]
I don't know, but it feels like its too much to implement such a logic of finding relevant documents, the different strategies of finding those on behalf of an editor. There definitely has to be a dependency inversion happening here.

However, I am unfamilliar with -cmp group of plugins as I barely use those.
Am I understanding it correct that you could do that 'depedency inversion' principle applying in a custom cmp plugin?

**Reply by McPatate** [reply]
> Am I understanding it correct that you could do that 'depedency inversion' principle applying in a custom cmp plugin?

I'm not sure I understand what you mean by that

**Reply by nenkoru** [reply]
> I'm not sure I understand what you mean by that

I meant that we could delegate the heavy-lifting of gathering of relevant documents to something else behind an abstraction layer.
In our case I proposed some kind of a rest api that we could send requests to, which would return relevant documents.
So that we send some context to it(filepath, current cursor position, etc.) and it process it with whatever algorithm or a strategy that layer uses. Either it is a vector search, or a Jaccard search, fzf, <you name it>.

Does nvim-cmp proposes any interfaces to interact with custom made things like that?

**Reply by McPatate** [reply]
Ah I see, yes it's definitely a good idea. I'm not aware of anything like that proposed by `nvim-cmp`. In neovim world the closest thing to that would be to create an LSP client.
I think we will eventually have to have a LSP server (well LSP or not, but it's just the easiest to integrate in IDEs) that does the heavy lifting

**Reply by nenkoru** [reply]
> Ah I see, yes it's definitely a good idea. I'm not aware of anything like that proposed by `nvim-cmp`. In neovim world the closest thing to that would be to create an LSP client.
> 
> I think we will eventually have to have a LSP server (well LSP or not, but it's just the easiest to integrate in IDEs) that does the heavy lifting

Got it. Do I understand it right that `-cmp` family of plugins provides the same functionality as this very repository, but also a neat way of choosing all the possible completions instead of only one?

**Reply by McPatate** [reply]
I'd tend to say yes, though there are some differences :)

**Reply by nenkoru** [reply]
> Ah I see, yes it's definitely a good idea. I'm not aware of anything like that proposed by `nvim-cmp`. In neovim world the closest thing to that would be to create an LSP client.
> 
> I think we will eventually have to have a LSP server (well LSP or not, but it's just the easiest to integrate in IDEs) that does the heavy lifting

Alright, after a little bit of digging I tend to choose LSP server. The protocol already defined, and we just need to implement it. And it also could be used not only in hfcc, but in general within IDEs.

I am more into Python development, however ready to try new languages to get them under the belt of mine. 

Do you think it would be fine to have an implementation in Python?

**Reply by nenkoru** [reply]
And if I understand it right, there's just a server needs to be implemented.

Specifically with completion capability with the logic of requesting HuggingFace model and completion. 

Not only completion, but also refactoring could be done much easier, as LSP already defines a way for that.

**Reply by McPatate** [reply]
> Do you think it would be fine to have an implementation in Python?

I would rather choose a language that compiles to cross-platform self-contained binaries (think no extra dependencies). We use Rust extensively at Hugging Face, so that'd be my language of choice in this case.

> there's just a server needs to be implemented

The server and the client must be updated as well, but yes you are correct :)

**Reply by McPatate** [reply]
I'll create the repo and setup the initial code when I have some time, is that ok with you? Do you have any time constraints on your side?

**Reply by nenkoru** [reply]
> The server and the client must be updated as well, but yes you are correct :)

Yeah, and integrate into the hfcc using [this piece of documentation from neovim](https://neovim.io/doc/user/lsp.html)

**Reply by nenkoru** [reply]
> I'll create the repo and setup the initial code when I have some time, is that ok with you? Do you have any time constraints on your side?

If you have any etas - yea but meanwhile I would try to tinker with small POC in python and change the client side accordingly.

**Reply by McPatate** [reply]
sgtm, I'll create a `lsp-poc` branch on which we can experiment and you can commit to. You can commit the lsp server code there as well for simplicity, I'll create a separate repo when you have a demo!

**Reply by McPatate** [reply]
https://github.com/huggingface/hfcc.nvim/tree/lsp-poc

**Reply by McPatate** [reply]
> If you have any eta

I'll initialise the repo this week

**Reply by nenkoru** [reply]
Alright, I made a little LSP server and tried to integrate it with nvim-cmp.
And it turns out there would be two paths:
1) [To trade-off inline suggestion](https://www.youtube.com/watch?v=wsqCrfPPxuM)
2) To implement the LSP client and bring inline-suggestion back

Also, what I noticed with the first way is that even if I execute a completion from `nvim-cmp` - it won't do a thing unless a dot(.) typed. Same happens in Sublime.
Because we brought a protocol, and we need to comply with it. We would not be using `nvim-cmp`, but `hfcc.nvim` itself would be a client which would bring that inline-suggestion back.

And as I see, almost all the editors and plugins which implement the LSP are all working ± the same. Meaning that they all do not have any kind of inline preview of the completion, but just in a preview window.

Personally, I would like to still have inline-suggestion because it feels more handy, because it automatically shows what's best could be inserted. Where `nvim-cmp` does not, it waits and then shows. 

At this point, I start to lean towards basic REST API server, which would have all the things we need. But it has its own shortcomings, like a proprietary to this specific project protocol.

Tell me what you think.

**Reply by McPatate** [reply]
I'm not sure I understand why you tried to use `nvim-cmp`. I think this should be a separate project altogether, I agree  I like the inline suggestions and I think that's what `hfcc.nvim` should be focusing on, regardless of using a language server or not (whether REST or LSP).

**Reply by McPatate** [reply]
Happy to see some code if you're willing to share!

**Reply by nenkoru** [reply]
> I'm not sure I understand why you tried to use `nvim-cmp`. I think this should be a separate project altogether, I agree I like the inline suggestions and I think that's what `hfcc.nvim` should be focusing on, regardless of using a language server or not (whether REST or LSP).

I believe I just gone a little bit off the track with `nvim-cmp` as I found it earlier. I thought it would be great to just create an LSP server which would do all the heavy lifting of requesting the model endpoint.

At this point I am totally unsure what exactly to put into LSP server.
Do we need to fully delegate the logic of completion to the server? If so, wouldn't it mean that `hfcc.nvim` would just become a thin-client for that server which kind of making it less independent as it was intended(was it?) in the first place?

Edited:
There is also a question left regarding which way exactly to implement:
1) Stdin(stdio)
2) TCP
3) Websockets

**Reply by nenkoru** [reply]
> Happy to see some code if you're willing to share!


[Here you are.](https://gist.github.com/nenkoru/ffd385707232f47890157724d9e65838)
It uses [pygls](https://github.com/openlawlibrary/pygls) so make sure to install it.
I was able to integrate it by making the script executable and putting the folder into PATH so that every editor and `nvim-cmp` in particular could access it. I used the default `nvim-cmp` example from a a README from a [repository](https://github.com/hrsh7th/nvim-cmp). I changed `'<YOUR_LSP_SERVER>'` to `pylsp` and after `capabilities = capabilities` I did `cmd={"hfcc_poc.py"}` so that it overrides the default command in lspconfig it uses under the hood to use `hfcc_poc.py` instead of `pylsp`

**Reply by McPatate** [reply]
I understand now why you dug into `nvim-cmp`.

I wonder if going the LSP route is what we want. We may need to create something custom based on REST or whatever protocol.

**Reply by McPatate** [reply]
There actually may be interesting things to do with a language server and would avoid, IIUC, developing a custom cmp plugin.

But hfcc.nvim in its current form still has value and will be difficult to merge with an LSP. I have to do some more research!

**Reply by nenkoru** [reply]
> There actually may be interesting things to do with a language server and would avoid, IIUC, developing a custom cmp plugin.
> 
> But hfcc.nvim in its current form still has value and will be difficult to merge with an LSP. I have to do some more research!

Well yeah, implenting a client within hfcc doesn't mean we need to go fully `-cmp` alike.
We would just be using that LSP as a provider for inline completions as hfcc does rn.
So that all the heay lifting is done within that LSP and hfcc just shows it as it does right now.

**Reply by nenkoru** [reply]
I actually start to think that implementing an LSP server would actually be quite handy. We could go with TCP way and use good old curl as it is doing right now. I vouch for LSP at this point.

Pros:
- Standard(no need to invent new protocols etc.)
- Flexibility(other IDEs could just integrate it and use it)

Cons:
- A little hard to implement

pygls particularly already has a [neat way of starting a TCP server](https://pygls.readthedocs.io/en/latest/pages/advanced_usage.html). I bet it would be easy to do so in Rust as well.
(just as a poc I could do it faster with pygls)

**Reply by nenkoru** [reply]
Okay, no.
Having LSP just puts a lot of coordination burden on our side.
For instance there is no way(at least as I see) right now to communicate `statelessly`. Without continuously maintaining a bi-directional connection between a client and a server. With all the ways of transports: TCP, Websockets, Stdio.
Which just make no sense for our use-case. I have tried searching for information regarding not maintaining a connection and just sending textComplete RPC call - and I was unable to find any info regarding that.
So at this point: RESTful API, or just a basic JSON-RPC server which complies with textComplete RPC protocol.

**Reply by nenkoru** [reply]
So, I believe I would try implementing a simple JSON-RPC via HTTP this weekend. And would try to integrate it.
I would offload all the logic of sending request to the model on the server side. And a procedure for the RPC call would be receiving a link(or a model name) into that procedure, as well maximum context length, etc.
Would reply with the final signature of a procedure here.
Update me if you have other thoughts on this.

**Reply by McPatate** [reply]
Ok, so I think we should try going the LSP route. To add to your list of pros, there are a number of things that the IDEs handle out of the box, such as in the case of neovim, methods defined in [vim.lsp](https://neovim.io/doc/user/lsp.html).
hfcc could then use the textDocument/completion, we should try it out and see if doesn't clash too much with regular autocomplete (cf request in https://neovim.io/doc/user/lsp.html#lsp-rpc).

**Reply by McPatate** [reply]
I'll push my rust server in a bit and link it here so we can collaborate on the project!

**Reply by McPatate** [reply]
Alright here it is : https://github.com/huggingface/ccserver

I've still yet to decide what method of the LSP protocol we'll be using for completions. As I said above, we can leverage `vim.lsp.rpc.request()` to send requests to the LSP in `hfcc`. Lots to do!

**Reply by nenkoru** [reply]
Alright, was quite busy this week. 
So what's the current status and what I could take to implement?

**Reply by McPatate** [reply]
No probs! Gave me some time to start hacking :)

I'm currently trying to be +/- "iso-feature", almost done, you can check https://github.com/huggingface/hfcc.nvim/pull/32 & https://github.com/huggingface/llm-ls/pull/2 to see where I'm at.

Once that's done, I was planning to start experimenting how to build a good prompt. So either looking at the vectordb route or other things. I'd suggest waiting for me to finish at least that first layer before starting coding on your side, I have some cleaning up to do 😄

**Reply by McPatate** [reply]
Some of the things I'd like to explore, in no particular order:
- building a good prompt
- request caching
- improving request latency
- run a model locally (thinking of using https://github.com/huggingface/candle)
- improve multiline / single line completions

**Reply by McPatate** [reply]
@nenkoru merged, let's get started!
